File: jdis-games-2023_starter-packs-main/python/network/network.py
import asyncio
import websockets

from core.game_state import GameState
from core.action import Action
from src.bot import MyBot

class Socket:
    def __init__(self, url: str, secret: str) -> None:
        self.__url = url
        self.__secret = secret
        self.__ws = None
        self.__queue = asyncio.Queue[GameState](maxsize=1)

        self.__bot = MyBot()

    async def __connect(self):
        extra_headers = { 'Authorization': self.__secret }
        try:
            async with websockets.connect(self.__url, extra_headers=extra_headers) as self.__ws:
            
                while True:
                    message = await self.__ws.recv()
                    state = GameState.deserialize(message)
                    try:
                        self.__queue.put_nowait(state)
                    except asyncio.QueueFull:
                        self.__queue.get_nowait()
                        self.__queue.put_nowait(state)

        except Exception as e:
            print(e)

    async def __process_queue(self):
        while True:
            state = await self.__queue.get()
            action = self.__bot.tick(state)
            await self.__ws.send(action.serialize())

    async def run(self):
        await asyncio.gather(self.__connect(), self.__process_queue())

File: jdis-games-2023_starter-packs-main/javascript/network/network.js
import { WebSocket } from 'ws';
import { GameState }  from '../core/game_state.js';
import { MyBot } from '../src/bot.js';

class Socket {
  constructor(url, secret) {
    this.url = url;
    this.secret = secret;
    this.ws = null;
    this.queue = [];

    this.bot = new MyBot();
    this.isProcessing = false;
  }

  async connect() {
    const ws = new WebSocket(this.url, {
      headers: { 'Authorization': this.secret }
    });
    ws.binaryType = 'arraybuffer';

    console.log(`Connected to ${this.url}`);

    ws.on('open', () => {});

    ws.on('message', async (message) => {
      const state = GameState.deserialize(message);
      try {
        this.queue.push(state);
        if (!this.isProcessing) {
          this.processQueue();
        }
      } catch(err) {
        console.error(err);
      }
    });

    this.ws = ws;
  }

  async processQueue() {
    this.isProcessing = true;
    while (this.queue.length > 0) {
      const state = this.queue.shift();
      const action = this.bot.tick(state);

      this.ws.send(action.serialize());
    }
    this.isProcessing = false;
  }

  async run() {
    await this.connect();
  }
}

export { Socket };
